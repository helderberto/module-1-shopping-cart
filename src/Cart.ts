import find from 'lodash/find';
import remove from 'lodash/remove';

export interface Product {
  title: string;
  price: number;
}

interface Item {
  quantity: number;
  product: Product;
}

export interface Bag {
  items: Array<Item>;
  getTotal(): number;
  add(item: Item): void;
  remove(product: Product): void;
  summary(): void;
  checkout(): void;
}

export class Cart implements Bag {
  items: Array<Item> = [];

  getTotal(): number {
    return this.items.reduce(
      (acc, item: Item) => acc + item.quantity * item.product.price,
      0,
    );
  }

  add(item: Item): void {
    const itemToMatch = { product: item.product };

    if (find(this.items, itemToMatch)) {
      remove(this.items, itemToMatch);
    }
    this.items.push(item);
  }

  remove(product: Product): void {
    return;
  }

  summary(): void {
    return;
  }

  checkout(): void {
    return;
  }
}
