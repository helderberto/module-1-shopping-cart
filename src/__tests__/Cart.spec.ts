import { Cart, Bag, Product } from '../Cart';

describe('Cart', () => {
  let cart: Bag;
  const product: Product = {
    title: 'Adidas running shoes - men',
    price: 35388,
  };

  const product2: Product = {
    title: 'Adidas running shoes - women',
    price: 40000,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal()).toBe(0);
    });

    it('should multiply quantity with price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2,
      };

      cart.add(item);

      expect(cart.getTotal()).toBe(70776);
    });

    it('should ensure no more than one product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal()).toBe(35388);
    });

    it('should update total when a product get included and then removed from cart', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal()).toBe(40000);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 1,
      });

      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.checkout()).toMatchInlineSnapshot(`
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 1,
    },
    Object {
      "product": Object {
        "price": 40000,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
  ],
  "total": 115388,
}
`);
    });
  });
});
