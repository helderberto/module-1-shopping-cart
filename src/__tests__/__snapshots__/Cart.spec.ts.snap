// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 40000,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
  ],
  "total": 150776,
}
`;
